### PR TITLE
[672] Modal Missing Focus Handler Just Inside Its Container

### PR DESCRIPTION
**Pull request by soujvnunes**
# [Modal Missing Focus Handler Just Inside Its Container](https://app.shortcut.com/polkamarkets/story/672/modal-missing-focus-handler-just-inside-its-container)

Hooks for **click away** and **trap focus** implemented inside a container, based on the referenced list.

The `useClickway` watches if the container loses its focus, instead of listen for a 'click outside', which is an [anti-pattern](https://reactjs.org/docs/accessibility.html#mouse-and-pointer-events).

For `useTrapfocus`, it is just focusing the first/end action elements when the container virtual trap-elements area focused. I guess it's faster than using listeners for _keyboard_ or _click_ events, since we would call methods everytime user clicks **Tab** key, while setting methods on focus event is declarative (I think so). And this navigates for and backwards!

I'm adding some xtra feats, such as:
1. _Escape_ key close modal;
2. Fixing `Checkbox` UI for focus feeback;
3. the focus returning to the previous DOM element before Modal was open 🔥

### References

1. https://muffinman.io/blog/catching-the-blur-event-on-an-element-and-its-children/
4. https://reactjs.org/docs/events.html#detecting-focus-entering-and-leaving
5. https://dev.to/mohitkyadav/how-to-trap-focus-in-react-3in8

https://user-images.githubusercontent.com/16447765/188677214-cbc0e54c-8a1f-4fae-a7a8-3fbaa4f7a4b8.mp4

### Test **click away**
1. Open a modal passing its hid function through `[onHide]`;
2. Click outside modal container. The modal shall close! Actually, it'll close even if you escape the page!
### Test **trap focus**
1. With **tab** key, navigate through modal! The focus shall not escape it! Even if you navigate backwards, it'll remain inside modal container.

## ⚡ Changelog

| _Unit_            | 🟢 Added                                                           | 🟡 Changed                                   | 🔴 Removed                  |
| ----------------- | ------------------------------------------------------------------ | -------------------------------------------- | --------------------------- |
| 🟢 useTrapfocus   | Traps focus inside a container throught its DOM node reference     |                                              |                             |
| 🟢 useClickaway   | Watches if a container loses focus throught its DOM node reference |                                              |                             |
| 🟡 eslintrc       |                                                                    | Extends jsx-a11y plugin                      |                             |
| 🟡 Modal          |                                                                    | Implements `useClickaway` and `useTrapfocus` |                             |
|                   |                                                                    | Calls `[onHide]` if user press Escape key    |                             |
|                   |                                                                    | Enhances constants data naming               |                             |
|                   |                                                                    | Removes unnecessary focus feedback           |                             |
| 🟡 Checkbox       |                                                                    | Fixes UI focus feedback                      |                             |
| 🔴 useEnhancedRef |                                                                    |                                              | `useClickaway` does the job |

## 🐛 Backlog

| _Unit_       | Has story? | Description         |
| ------------ | ---------- | ------------------- |
| useTrapfocus | 🔴         | Missing unity tests |
| useClickaway | 🔴         | Missing unity tests |

### Footnotes

> `useHook()` is related to React JS hooks, located on `src/hooks/` path;
> 
> `<Component />`/ `<Component [prop=value] />` is related to React JS components, located on `src/components/` path;
> 
> `module()` is related to modules located on `src/utils/` or anywhere else on the app that provides some usefull shared resource;

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -2,12 +2,9 @@
   "compilerOptions": {
     "incremental": true,
     "outDir": "build/lib",
-    "target": "es5",
+    "target": "ES6",
```
Comment (soujvnunes) [inline tsconfig.json:5]: I need to upgrade this from `es5` and now we're getting the eslint warnings twice (an example):

image

May some help? I can revert this, but I'll have to interate between list of nodes myself... ([src/hooks/useEnhancedRef/useEnhancedRef.ts](https://github.com/Polkamarkets/polkamarkets-web/pull/150/files#diff-7bdc04b19aab1114fe7a88ea760d83f6b065cbf949c8f3ed71e1fdf78f156e26)#31)

File: src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -63,11 +64,15 @@ function Modal({ children, onHide, show, name }: ModalProps) {
...
                 role="dialog"
                 aria-modal="true"
-                aria-labelledby={`${name}-${MODAL_DATA.title}`}
-                aria-describedby={`${name}-${MODAL_DATA.description}`}
+                aria-labelledby={`${name}-${modalData.title}`}
+                aria-describedby={`${name}-${modalData.description}`}
                 className="pm-c-modal"
               >
+                {/** todo: to be virtually created */}
+                <span tabIndex={0} id="trap-start" />
```
Comment (soujvnunes) [inline src/components/Modal/Modal.tsx:72]: It’s to be created through virtualization by the `useFocustrap`

File: src/react-app.d.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import * as React from 'react';
+
+declare module 'react' {
+  export interface PropsWithChildrenElement {
```
Comment (soujvnunes) [inline src/react-app.d.ts:4]: this is not used. Remove it

File: src/hooks/useClickaway/useClickaway.type.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export type useClickawayProps = {
```
Comment (soujvnunes) [inline src/hooks/useClickaway/useClickaway.type.ts:1]: This is not used. Remove it

**Comment by soujvnunes** [comment]
lock files will be removed

File: src/components/Modal/Modal.tsx [hunk 5]
```diff
@@ -43,6 +39,14 @@ function Modal({ children, onHide, show, name }: ModalProps) {
...
     }
   }, [Portal, show, showPrev]);
+  useClickaway(ref, () => {
+    if (showPrev && show) onHide?.();
+  });
+  useTrapfocus(ref);
+
+  function handleKeyDown(event: React.KeyboardEvent) {
+    if (event.key === 'Escape') onHide?.();
+  }
```
Comment (soujvnunes) [inline src/components/Modal/Modal.tsx:49]: This is the simpliest way to implement this. In the future, if we use it anywhere else, we may extract then

**Review by WellingtonMatheus — APPROVED** [review]

**Pull request merged**